# Bank Marketing Classification

## Machine Learning Capstone Project

### Classification Track - Review 1

This notebook implements and evaluates machine learning classification algorithms using the Bank Marketing dataset.

The objective is to predict whether the client will subscribe to a term deposit on demographic, financial, and previous marketing campaign information.

This notebook covers: 

1. Dataset Loading and Audit
2. Exploratory Data Analysis
3. Data Cleaning
4. Preprocessing and Feature Engineering
5. Classification Model Implementation 
6. Model Evaluation and Comparison

# 1. Importing Required Libraries

In [4]:
# Data Manipulation
import pandas as pd
import numpy as np
# Data Visualization
import matplotlib.pyplot as plt
import seaborn as sns
# Display Settings
pd.set_option('display.max_columns', None)
# Reproductibility
RANDOM_STATE = 42

# 2. Dataset Loading 

For the initial pahse of this project, we are use `bank.csv`, smaller subset of the complete Bank Marketing dataset. This allows us to develop and test the machine learning pipeline before later comparing the results with the complete dataset.

In [5]:
# Load the dataset
file_path = '../data/classification/bank.csv'
df = pd.read_csv(file_path, sep=';')
df.head()

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,30,unemployed,married,primary,no,1787,no,no,cellular,19,oct,79,1,-1,0,unknown,no
1,33,services,married,secondary,no,4789,yes,yes,cellular,11,may,220,1,339,4,failure,no
2,35,management,single,tertiary,no,1350,yes,no,cellular,16,apr,185,1,330,1,failure,no
3,30,management,married,tertiary,no,1476,yes,yes,unknown,3,jun,199,4,-1,0,unknown,no
4,59,blue-collar,married,secondary,no,0,yes,no,unknown,5,may,226,1,-1,0,unknown,no


# 3.Dataset Audit

This section examines the structure and quality of the dataset, including its dimensions, data types, missing values, and target class distribution.

In [6]:
# Dataset dimensions
print("Dataset shape:", df.shape)
# Column data types
print("\n Data types:")
display(df.dtypes)
# Missing values
print("\n Missing values:")
display(df.isnull().sum())
# Target class distribution
print("\n Target class distribution:")
display(df['y'].value_counts())
print("\n Target class distribution (percentage):")
display(df['y'].value_counts(normalize=True) * 100)

Dataset shape: (4521, 17)

 Data types:


age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object


 Missing values:


age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64


 Target class distribution:


y
no     4000
yes     521
Name: count, dtype: int64


 Target class distribution (percentage):


y
no     88.476001
yes    11.523999
Name: proportion, dtype: float64

### Dataset Audit Insights 

- The dataset contains 4,521 records, consisting of 16 input features and one target variable (`y`).
- The dataset contains both numerical and categorical features, meaning categorical variables will require encoding before machine learning models can be trained.
- No missing values were identified in any column; therefore, missing-value imputation is not required.
- The target variable is significantly imbalanced, with approximately 88.48% of customers belonging to the `no` class and 11.52% belonging to the `yes` class.
- Because of this class imbalance, accuracy alone may not provide a complete evaluation of model performance. Additional evaluation metrics will therefore be considered.